### PR TITLE
[#84] Support saving and loading

### PR DESCRIPTION
**Pull request by hangpark**
See #84.

# :congratulations: Congratulations for the :100:th iss/pr :congratulations:

### CODE REVIEW

**Review by bimak-server — APPROVED** [review]
👍 

㊗️ Congratulations for the 💯th iss/pr ㊗️

File: app/src/main/java/kr/ac/kaist/team888/mylittlefont/FontItem.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+  private int id;
+  private String name;
+  private Date datetime;
+  private double curve;
+  private double roundness;
+  private double weight;
+  private double width;
+
+  private FontItem(double curve, double roundness, double weight, double width) {
```
Comment (xeoh) [inline app/src/main/java/kr/ac/kaist/team888/mylittlefont/FontItem.java]: Why do we need this constructor?

File: app/src/main/java/kr/ac/kaist/team888/mylittlefont/MainActivity.java [hunk 2]
```diff
@@ -179,4 +187,83 @@ public void onClick(DialogInterface dialog, int which) {
...
+          @Override
+          public void onClick(View view) {
+            String name = nameInput.getText().toString();
+            // Name should not be empty.
+            if (name.isEmpty()) {
+              Toast.makeText(getApplicationContext(), R.string.feature_save_empty_name,
+                  Toast.LENGTH_SHORT).show();
+              return;
+            }
+            new DatabaseOpenHelper(getApplicationContext())
```
Comment (xeoh) [inline app/src/main/java/kr/ac/kaist/team888/mylittlefont/MainActivity.java:223]: Why don't we hold DBhelper in MainActivity? Loading multiple helper object every time we save seems inefficient.

File: app/src/main/java/kr/ac/kaist/team888/util/DatabaseOpenHelper.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  private static final String KEY_DATETIME = "date";
+  private static final String KEY_CURVE = "curve";
+  private static final String KEY_ROUNDNESS = "roundness";
+  private static final String KEY_WEIGHT = "weight";
+  private static final String KEY_WIDTH = "width";
+
+  private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+
+  private static final String TABLE_CREATE =
+      "CREATE TABLE " + TABLE_NAME + " ("
```
Comment (xeoh) [inline app/src/main/java/kr/ac/kaist/team888/util/DatabaseOpenHelper.java:35]: Hard to read type of each attribute.

```java
"CREATE TABLE " + TABLE_NAME + " ("
            + KEY_ID        + " INTEGER PRIMARY KEY AUTOINCREMENT, "
            + KEY_NAME      + " TEXT, "
            + KEY_DATETIME  + " DATETIME DEFAULT CURRENT_TIMESTAMP, "
            + KEY_CURVE     + " REAL, "
            + KEY_ROUNDNESS + " REAL, "
            + KEY_WEIGHT    + " REAL, "
            + KEY_WIDTH     + " REAL);";
```

File: app/src/main/java/kr/ac/kaist/team888/util/DatabaseOpenHelper.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    return results;
+  }
+
+  /**
+   * Delete the font data corresponding to the given id from the database.
+   *
+   * @param id id of the font data
+   */
+  public void deleteItem(int id) {
+    getWritableDatabase().execSQL(ROW_DELETE + id);
```
Comment (xeoh) [inline app/src/main/java/kr/ac/kaist/team888/util/DatabaseOpenHelper.java]:
```java
String ROW_DELETE = "DELETE FROM %s WHERE %d = %d";
execSQL(String.format(ROW_DELETE, TABLE_NAME, KEY_ID, id));
```

**Review by xeoh — CHANGES_REQUESTED** [review]

**Comment by xeoh** [comment]
Any way congrat for 💯th iss/pr

**Review by xeoh — APPROVED** [review]
👍

**Pull request merged**